include/libsemigroups/schreier-sims-helper.hpp

```diff
@@ -0,0 +1,182 @@
+// libsemigroups - C++ library for semigroups and monoids
+// Copyright (C) 2021 Reinis Cirpons
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// This file contains an implementation of a backtrack search for computing
+// the intersection of two permutations given by Schreier-Sims algorithm.
+
+// TODO:
+// 1. Implement orbit refinement heuristic for intersection.
+// 2. Make the Screier-Sims object during runtime, since we compute the
+//    stabilizers of the intersection already.
+// 3. Refactor for more generality.
+
+#ifndef LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+#define LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+
+#include <cstddef>  // for size_t
+
+#include "containers.hpp"     // for Array2
+#include "schreier-sims.hpp"  // for SchreierSims, SchreierSims<>::ele...
+#include "transf.hpp"         // for LeastPerm
+#include "types.hpp"          // for SmallestInteger
+
+#include <iostream>
+
+namespace libsemigroups {
+
+  namespace schreier_sims_helper {
+
+    template <size_t N>
+    using point_type = typename SchreierSims<N>::point_type;
+
+    template <size_t N>
+    using element_type = typename SchreierSims<N>::element_type;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using One = typename SchreierSims<N>::One;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using Product = typename SchreierSims<N>::Product;
+
+    //! Find the intersection of two permutation groups.
+    //!
+    //! Modifies the first parameter \p T to be the Schreier-Sims object
+    //! corresponding to the intersection of \p S1 and \p S2.
+    //!
+    //! \tparam N the largest point not fixed by the permutations in the
+    //! permutation groups.
+    //!
+    //! \param T the Screier-Sims object holding the result.
```

Comment:
```suggestion
    //! \param T the Schreier-Sims object holding the result.
```

Reply:
Also as a separate comment: what pre-conditions exist for `T`? Does `T` have to be empty? If so, this should be written in the doc, and checked in the code.

Reply:
Has to be empty. Added check to code and documentation.